slapo/op/fused_bias.py

```diff
@@ -118,3 +109,52 @@ def reset_parameters(self, prev_weight=None):
 
     def forward(self, inp):
         return self.func(inp, self.bias)
+
+
+def bias_dropout(
+    x: torch.Tensor,
+    bias: torch.Tensor,
+    p: float = 0.5,
+    training: bool = True,
+    inplace: bool = False,
+) -> torch.Tensor:
+    return F.dropout(x + bias, p=p, training=training, inplace=inplace)
+
+
+class FusedBiasDropout(torch.nn.Module):
+    def __init__(
+        self,
+        size,
+        p=0.5,
+        training=True,
+        inplace=False,
+        device=None,
+        dtype=None,
+        prev_weight=None,
+    ):
+        factory_kwargs = {"device": device, "dtype": dtype}
+        super().__init__()
+        self.bias = torch.nn.Parameter(torch.empty(size, **factory_kwargs))
+        self.reset_parameters(prev_weight)
+        self.p = p
+        self.training = training
+        self.inplace = inplace
+        self.func = torch.jit.script(bias_dropout)
+        # Somehow memory_efficient_fusion generates a different dropout mask
+        # against the original dropout function even the random seed is the same.
+        # if memory_efficient_fusion is not None:
+        #     bias_dropout_func = partial(
+        #         bias_dropout, p=p, training=training, inplace=inplace
+        #     )
+        #     self.func = memory_efficient_fusion(bias_dropout_func)
```

Comment:
what is this memory_efficient_fusion?

Reply:
This is a new API used in PyTorch 2.0. It applies AOTAutograd in advance (https://pytorch.org/functorch/stable/generated/functorch.compile.memory_efficient_fusion.html). The performance should be similar to TorchScript in our use cases.

Reply:
"partitioner to perform efficient recomputation" means activation recomputation?

Reply:
It should refer to their algorithm (https://dev-discuss.pytorch.org/t/467) that is similar to our rematerialization pass in RAF. It uses a min-cut algorithm to selectively recompute some tensors. However, again since we only apply this API to a small module, it should not make any impact.

Reply:
Yes. `memory_efficient_fusion` is a new API and should be getting mature in PT 2.0, so we may consider switching to that in the future.